terraform/context_components.go

```diff
@@ -46,10 +46,10 @@ func (c *basicComponentFactory) ResourceProvisioners() []string {
 	return result
 }
 
-func (c *basicComponentFactory) ResourceProvider(typ string) (providers.Interface, error) {
-	f, ok := c.providers[addrs.NewLegacyProvider(typ)]
+func (c *basicComponentFactory) ResourceProvider(typ addrs.Provider) (providers.Interface, error) {
+	f, ok := c.providers[typ]
 	if !ok {
-		return nil, fmt.Errorf("unknown provider %q", typ)
+		return nil, fmt.Errorf("unknown provider %q", typ.LegacyString())
```

Comment:
This seems like it should be just `typ.String()` because we wouldn't want to panic here if we're given a non-existing and _non-legacy_ provider. :thinking: